### ISSUE TITLE
`builtins.toDir`

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nix#3497]
We have `toFile`, `readFile, and `readDir`, this would complete the square.

-------

This probably can be solved another way in the meantime, but I would use it move the setup hooks common to every stdenv to `$out/nix-support/setup/hook`, that we the awful part of stdenv that tries to `source` an input if it is just a single file can go away.

 - Bad stdenv, bad: https://github.com/NixOS/nixpkgs/blob/2ef110ae64a96c0d16df86f8390cfe0af7ff45dc/pkgs/stdenv/generic/setup.sh#L474-L476

 - Bare setup hooks: https://github.com/NixOS/nixpkgs/blob/2a19a034313e46134853a655121d45b37ab3cbb6/pkgs/stdenv/generic/default.nix#L55-L69

**Comment by edolstra** [comment]
How does it differ from `builtins.path`?

**Comment by Ericson2314** [comment]
Let me write an example of what I am thinking:
```nix
builtins.toDir "nix-support" {
  "setup-hook" = { 
      executable = false;
      contents = ./pkgs/buils-suport/setup-hooks/foo.sh;
   };
}
```

`builtins.path` can't do that, right?

**Comment by roberth** [comment]
Smells like a lazy tree.

**Comment by edolstra** [comment]
Yeah, this could be implemented as an `InputAccessor`, similar to `PatchingInputAccessor` , that would synthesize a source tree lazily.

**Comment by adrian-gierakowski** [comment]
I want this

**Comment by roberth** [comment]
- Related: https://github.com/NixOS/nix/issues/14011
  - difference: a derivation so that it can reference other outputs; not just evaluated paths
  - difference: only a file, but that could of course be generalized

**Comment by roberth** [comment]
A primop like this may well perform better as a backend for Nixpkgs `lib.fileset`, because it wouldn't have to perform callbacks for each and every file and directory.
